File: fe2-android/app/src/test/java/com/github/dedis/popstellar/model/network/method/message/data/consensus/ConsensusPrepareTest.java
package com.github.dedis.popstellar.model.network.method.message.data.consensus;

import com.github.dedis.popstellar.model.network.JsonTestUtils;
import com.github.dedis.popstellar.model.network.method.message.data.Action;
import com.github.dedis.popstellar.model.network.method.message.data.Objects;
import com.github.dedis.popstellar.model.objects.security.MessageID;
import com.github.dedis.popstellar.testutils.Base64DataUtils;
import com.google.gson.JsonParseException;

import org.junit.Test;

import static org.junit.Assert.*;

public class ConsensusPrepareTest {

  private static final String instanceId = "aaa";
  private static final MessageID messageId = new MessageID("TVNHX0lE");
  private static final long timeInSeconds = 1635277619;
  private static final int proposedTry = 4;

  private static final ConsensusPrepare prepare =
      new ConsensusPrepare(instanceId, messageId, timeInSeconds, proposedTry);

  @Test
  public void getObjectTest() {
    assertEquals(Objects.CONSENSUS.getObject(), prepare.getObject());
  }

  @Test
  public void getActionTest() {
    assertEquals(Action.PREPARE.getAction(), prepare.getAction());
  }

  @Test
  public void getInstanceIdTest() {
    assertEquals(instanceId, prepare.getInstanceId());
  }

  @Test
  public void getMessageIdTest() {
    assertEquals(messageId, prepare.getMessageId());
  }

  @Test
  public void getCreationTest() {
    assertEquals(timeInSeconds, prepare.getCreation());
  }

  @Test
  public void getPrepareValueTest() {
    PrepareValue value = prepare.getPrepareValue();

    assertEquals(proposedTry, value.getProposedTry());

    PrepareValue value2 = new PrepareValue(proposedTry);
    assertEquals(value, value2);
    assertEquals(value.hashCode(), value2.hashCode());

    assertNotEquals(value, null);
    assertNotEquals(value, new PrepareValue(proposedTry + 1));
  }

  @Test
  public void equalsTest() {
    ConsensusPrepare prepare2 =
        new ConsensusPrepare(instanceId, messageId, timeInSeconds, proposedTry);
    assertEquals(prepare, prepare2);
    assertEquals(prepare.hashCode(), prepare2.hashCode());

    String random = "random";
    assertNotEquals(prepare, null);
    assertNotEquals(prepare, new ConsensusPrepare(random, messageId, timeInSeconds, proposedTry));
    assertNotEquals(
        prepare,
        new ConsensusPrepare(
            instanceId,
            Base64DataUtils.generateMessageIDOtherThan(messageId),
            timeInSeconds,
            proposedTry));
    assertNotEquals(
        prepare, new ConsensusPrepare(instanceId, messageId, timeInSeconds + 1, proposedTry));
    assertNotEquals(
        prepare, new ConsensusPrepare(instanceId, messageId, timeInSeconds, proposedTry + 1));
  }

  @Test
  public void toStringTest() {
    assertEquals(
        "ConsensusPrepare{instance_id='aaa', message_id='TVNHX0lE', created_at=1635277619, value=PrepareValue{proposed_try=4}}",
        prepare.toString());
  }

  @Test
  public void jsonValidationTest() {
    JsonTestUtils.testData(prepare);

    String dir = "protocol/examples/messageData/consensus_prepare/";
    String jsonInvalid1 = JsonTestUtils.loadFile(dir + "wrong_prepare_negative_created_at.json");
    String jsonInvalid2 = JsonTestUtils.loadFile(dir + "wrong_prepare_negative_proposed_try.json");
    assertThrows(JsonParseException.class, () -> JsonTestUtils.parse(jsonInvalid1));
    assertThrows(JsonParseException.class, () -> JsonTestUtils.parse(jsonInvalid2));
  }
}

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/consensus/ConsensusPrepareTest.kt
package com.github.dedis.popstellar.model.network.method.message.data.consensus

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.github.dedis.popstellar.model.network.JsonTestUtils
import com.github.dedis.popstellar.model.network.method.message.data.Action
import com.github.dedis.popstellar.model.network.method.message.data.Objects
import com.github.dedis.popstellar.model.objects.security.MessageID
import com.github.dedis.popstellar.testutils.Base64DataUtils
import com.google.gson.JsonParseException
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class ConsensusPrepareTest {

  private val instanceId = "aaa"
  private val messageId = MessageID("TVNHX0lE")
  private val timeInSeconds: Long = 1635277619
  private val proposedTry = 4
  private val prepare = ConsensusPrepare(instanceId, messageId, timeInSeconds, proposedTry)

  @Test
  fun objectTest() {
    Assert.assertEquals(Objects.CONSENSUS.getObject(), prepare.getObject())
  }

  @Test
  fun actionTest() {
    Assert.assertEquals(Action.PREPARE.action, prepare.action)
  }

  @Test
  fun instanceIdTest() {
    Assert.assertEquals(instanceId, prepare.instanceId)
  }

  @Test
  fun messageIdTest() {
    Assert.assertEquals(messageId, prepare.messageId)
  }

  @Test
  fun creationTest() {
    Assert.assertEquals(timeInSeconds, prepare.creation)
  }

  @Test
  fun prepareValueTest() {
    val value = prepare.prepareValue

    Assert.assertEquals(proposedTry.toLong(), value.proposedTry.toLong())

    val value2 = PrepareValue(proposedTry)

    Assert.assertEquals(value, value2)
    Assert.assertEquals(value.hashCode().toLong(), value2.hashCode().toLong())

    Assert.assertNotEquals(value, null)
    Assert.assertNotEquals(value, PrepareValue(proposedTry + 1))
  }

  @Test
  fun equalsTest() {
    val prepare2 = ConsensusPrepare(instanceId, messageId, timeInSeconds, proposedTry)

    Assert.assertEquals(prepare, prepare2)
    Assert.assertEquals(prepare.hashCode().toLong(), prepare2.hashCode().toLong())

    val random = "random"

    Assert.assertNotEquals(prepare, null)
    Assert.assertNotEquals(prepare, ConsensusPrepare(random, messageId, timeInSeconds, proposedTry))
    Assert.assertNotEquals(
      prepare,
      ConsensusPrepare(
        instanceId,
        Base64DataUtils.generateMessageIDOtherThan(messageId),
        timeInSeconds,
        proposedTry
      )
    )
    Assert.assertNotEquals(
      prepare,
      ConsensusPrepare(instanceId, messageId, timeInSeconds + 1, proposedTry)
    )
    Assert.assertNotEquals(
      prepare,
      ConsensusPrepare(instanceId, messageId, timeInSeconds, proposedTry + 1)
    )
  }

  @Test
  fun toStringTest() {
    Assert.assertEquals(
      "ConsensusPrepare{instance_id='aaa', message_id='TVNHX0lE', created_at=1635277619, value=PrepareValue{proposed_try=4}}",
      prepare.toString()
    )
  }

  @Test
  fun jsonValidationTest() {
    JsonTestUtils.testData(prepare)

    val dir = "protocol/examples/messageData/consensus_prepare/"
    val jsonInvalid1 = JsonTestUtils.loadFile(dir + "wrong_prepare_negative_created_at.json")
    val jsonInvalid2 = JsonTestUtils.loadFile(dir + "wrong_prepare_negative_proposed_try.json")

    Assert.assertThrows(JsonParseException::class.java) { JsonTestUtils.parse(jsonInvalid1) }
    Assert.assertThrows(JsonParseException::class.java) { JsonTestUtils.parse(jsonInvalid2) }
  }
}
